u2f-ref-code/java/src/com/google/u2f/server/impl/U2FServerReferenceImpl.java

```diff
@@ -267,6 +290,99 @@ public SecurityKeyData processSignResponse(SignResponse signResponse) throws U2F
+   *       bluetoothLowEnergyRadio(1),
+   *       uSB(2),
+   *       nFC(3)
+   *     }
+   *
+   *   <p>Note that the BIT STRING must be wrapped in an OCTET STRING.
+   *   An extension that encodes BT, BLE, and NFC then looks as follows:
+   *
+   *   <p>SEQUENCE (2 elem)
+   *      OBJECT IDENTIFIER 1.3.6.1.4.1.45724.2.1.1
+   *      OCTET STRING (1 elem)
+   *        BIT STRING (4 bits) 1101
+   *
+   * @param cert the certificate to parse for extension
+   * @return the supported transports as a List of HardwareFeatures or null if no extension
+   * was found
+   */
+  public static List<Transports> parseTransportsExtension(X509Certificate cert)
+      throws CertificateParsingException{
+    byte[] extValue = cert.getExtensionValue(TRANSPORT_EXTENSION_OID);
+    LinkedList<Transports> transportsList = new LinkedList<Transports>();
+    if (extValue == null) {
+      // No transports extension found.
+      return null;
+    }
+
+    ASN1InputStream ais = new ASN1InputStream(extValue);
+    ASN1Object asn1Object;
+    // Read out the OctetString
+    try {
+      asn1Object = ais.readObject();
+      ais.close();
+    } catch (IOException e) {
+      throw new CertificateParsingException("Not able to read object in transports extenion", e);
+    }
+
+    if (asn1Object == null || !(asn1Object instanceof DEROctetString)) {
+      throw new CertificateParsingException("No Octet String found in transports extension");
+    }
+    DEROctetString octet = (DEROctetString) asn1Object;
+
+    // Read out the BitString
+    ais = new  ASN1InputStream(octet.getOctets());
+    try {
+      asn1Object = ais.readObject();
+      ais.close();
+    } catch (IOException e) {
+      throw new CertificateParsingException("Not able to read object in transports extension", e);
+    }
+    if (asn1Object == null || !(asn1Object instanceof DERBitString)) {
+      throw new CertificateParsingException("No BitString found in transports extension");
+    }
+    DERBitString bitString = (DERBitString) asn1Object;
+
+    byte [] values = bitString.getBytes();
+    int nrBytesUsed = values.length;
+    // Check if more than one byte was used to store the transport values
+    // If so, we need to reverse the byte order before feeding them to a BitSet object
+    // This is because when reading from a byte array, a BitSet starts reading from the end
+    if (nrBytesUsed > 1) {
+      values = reverseOrderOfBytes(values);
+    }
+    BitSet bitSet = BitSet.valueOf(values);
+
+    // Parse the actual transport values
+    int nrBitsUsed = nrBytesUsed * BITS_IN_A_BYTE;
+    // We might have more defined transports than used by the extension
+    for (int i = 0; i < Transports.values().length && i < nrBitsUsed; i++) {
+      if (bitSet.get(nrBitsUsed - i - 1)) {
+        transportsList.add(Transports.values()[i]);
+      }
+    }
+    return transportsList;
+  }
+
+  private static byte [] reverseOrderOfBytes (byte[] input) {
+    byte [] output = new byte[input.length];
```

Comment:
Thanks for the pointer. I've been looking for a library to do the reversing too and hadn't found one in the standard Java library.

I had a try at Collections.reverse and unfortunately ended up with more lines of code than before. Converting from byte[] to Byte[] and back seems to make things messy. I'd leave the code as is, even though I am not a fan of reversing bytes by hand and keeping track of indices +-1, etc. I think this method is small enough to be fine as is. 

Please do let me know if you see a cleaner way of doing the reversal than the blob I'm pasting here. I'd be happy to improve this code.

  private static byte [] reverseOrderOfBytes (byte[] input) {
   // Arrays.asList(input) would consider input the Object and give me back a list with
 // one element in it, my byte[]. So I am creating the list by hand :-/
    List<Byte> toReverse = new LinkedList();
    for (int i = 0; i < input.length; i++) {
       toReverse.add((Byte) input[i]);
    }
   Collections.reverse(toReverse);
   Byte [] reversed = new Byte[input.length];
   reversed = toReverse.toArray(reversed);
  // now I got a Byte[] of the reversed values, but i have to manually convert it to a byte[]. No lib that   
  // I know of to do this 
   byte[] output = new byte[input.length];
   for (int i = 0; i < output.length; i++) {
     output [i] = reversed[i];  
  }
  return output;
  }